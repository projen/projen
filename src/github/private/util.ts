export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

export function context(value: string) {
  return `\${{ ${value} }}`;
}

// Checks if part of the file path is hidden
export function isHiddenPath(path: string) {
  return /(^|\/)\.[^\/\.]/g.test(path);
}

// Helper to assert a path is not hidden
export function ensureNotHiddenPath(value: string, name: string) {
  if (isHiddenPath(value)) {
    throw Error(`${name} cannot be a hidden path, got: ${value}`);
  }
}
