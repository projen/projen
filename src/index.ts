export * from "./ai-instructions";
export * from "./compare";
export * from "./component";
export * from "./dev-env";
export * from "./dependencies";
export * from "./docker-compose";
export * from "./dockerfile";
export * from "./file";
export * from "./gitattributes";
export * from "./gitpod";
export * from "./runner-options";
export * from "./ignore-file";
export * from "./ini";
export * from "./json";
export * from "./json-patch";
export * from "./logger";
export * from "./license";
export * from "./makefile";
export * from "./object-file";
export * from "./option-hints";
export * from "./project";
export * from "./project-build";
export * from "./project-tree";
export * from "./projects";
export * from "./projenrc";
export * from "./projenrc-json";
export * from "./readme";
export * from "./renovatebot";
export * from "./sample-file";
export * from "./semver";
export * from "./source-code";
export * from "./task";
export * from "./tasks";
export * from "./task-model";
export * from "./task-runtime";
export * from "./testing";
export * from "./textfile";
export * from "./toml";
export * from "./version";
export * from "./yaml";
export * from "./xmlfile";

// export submodules
export * as web from "./web";
export * as github from "./github";
export * as gitlab from "./gitlab";
export * as circleci from "./circleci";
export * as vscode from "./vscode";
export * as java from "./java";
export * as python from "./python";
export * as typescript from "./typescript";
export * as javascript from "./javascript";
export * as release from "./release";
export * as awscdk from "./awscdk";
export * as cdk8s from "./cdk8s";
export * as cdktf from "./cdktf";
export * as cdk from "./cdk";
export * as build from "./build";
