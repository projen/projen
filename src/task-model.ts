/**
 * Schema for `tasks.json`.
 */
export interface TasksManifest {
  /**
   * All tasks available for this project.
   */
  readonly tasks?: { [name: string]: TaskSpec };

  /**
   * Environment for all tasks.
   */
  readonly env?: { [name: string]: string };
}

export interface TaskCommonOptions {
  /**
   * The description of this build command.
   * @default - the task name
   */
  readonly description?: string;

  /**
   * Defines environment variables for the execution of this task.
   * Values in this map will be evaluated in a shell, so you can do stuff like `$(echo "foo")`.
   * @default {}
   */
  readonly env?: { [name: string]: string };

  /**
   * A set of environment variables that must be defined in order to execute
   * this task. Task execution will fail if one of these is not defined.
   */
  readonly requiredEnv?: string[];

  /**
   * A shell command which determines if the this task should be executed. If
   * the program exits with a zero exit code, steps will be executed. A non-zero
   * code means that task will be skipped.
   */
  readonly condition?: string;

  /**
   * The working directory for all steps in this task (unless overridden by the
   * step).
   *
   * @default - process.cwd()
   */
  readonly cwd?: string;
}

/**
 * Specification of a single task.
 */
export interface TaskSpec extends TaskCommonOptions {
  /**
   * Task name.
   */
  readonly name: string;

  /**
   * Task steps.
   */
  readonly steps?: TaskStep[];
}

/**
 * Options for task steps.
 */
export interface TaskStepOptions {
  /**
   * Step name
   *
   * @default - no name
   */
  readonly name?: string;

  /**
   * The working directory for this step.
   *
   * @default - determined by the task
   */
  readonly cwd?: string;

  /**
   * A shell command which determines if the this step should be executed. If
   * the program exits with a zero exit code, the step will be executed. A non-zero
   * code means the step will be skipped (subsequent task steps will still be evaluated/executed).
   */
  readonly condition?: string;

  /**
   * Should this step receive args passed to the task.
   *
   * If `true`, args are passed through at the end of the `exec` shell command.\
   * The position of the args can be changed by including the marker `$@` inside the command string.
   *
   * If the marker is explicitly double-quoted ("$@") arguments will be wrapped in single quotes, approximating
   * the whitespace preserving behavior of bash variable expansion.
   *
   * If the step spawns a subtask, args are passed to the subtask.
   * The subtask must define steps receiving args for this to have any effect.
   *
   * @example task.exec("echo Hello $@ World!", { receiveArgs: true });
   *
   * @default false
   */
  readonly receiveArgs?: boolean;

  /**
   * A list of fixed arguments always passed to the step.
   *
   * Useful to re-use existing tasks without having to re-define the whole task.\
   * Fixed args are always passed to the step, even if `receiveArgs` is `false`
   * and are always passed before any args the task is called with.
   *
   * If the step executes a shell commands, args are passed through at the end of the `exec` shell command.\
   * The position of the args can be changed by including the marker `$@` inside the command string.
   *
   * If the step spawns a subtask, args are passed to the subtask.
   * The subtask must define steps receiving args for this to have any effect.
   *
   * If the step calls a builtin script, args are passed to the script.
   * It is up to the script to use or discard the arguments.
   *
   * @example task.spawn("deploy", { args: ["--force"] });
   *
   * @default - no arguments are passed to the step
   */
  readonly args?: string[];

  /**
   * Defines environment variables for the execution of this step (`exec` and `builtin` only).
   * Values in this map can be simple, literal values or shell expressions that will be evaluated at runtime e.g. `$(echo "foo")`.
   *
   * @example { "foo": "bar", "boo": "$(echo baz)" }
   *
   * @default - no environment variables defined in step
   */
  readonly env?: { [name: string]: string };
}

/**
 * A single step within a task. The step could either be  the execution of a
 * shell command or execution of a sub-task, by name.
 */
export interface TaskStep extends TaskStepOptions {
  /**
   * Shell command to execute
   *
   * @default - don't execute a shell command
   */
  readonly exec?: string;

  /**
   * Subtask to execute
   *
   * @default - don't spawn a subtask
   */
  readonly spawn?: string;

  /**
   * Print a message.
   * @default - don't say anything
   */
  readonly say?: string;

  /**
   * The name of a built-in task to execute.
   *
   * Built-in tasks are node.js programs baked into the projen module and as
   * component runtime helpers.
   *
   * The name is a path relative to the projen lib/ directory (without the .task.js extension).
   * For example, if your built in builtin task is under `src/release/resolve-version.task.ts`,
   * then this would be `release/resolve-version`.
   *
   * @default - do not execute a builtin task
   */
  readonly builtin?: string;
}
