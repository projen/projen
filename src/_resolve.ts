import { types } from "util";
import { ResolveOptions, IResolvable } from "./file";

function isResolvable(obj: any): obj is IResolvable {
  return (obj as IResolvable).toJSON !== undefined;
}

export function resolve(value: any, options: ResolveOptions = {}): any {
  const args = options.args ?? [];
  const omitEmpty = options.omitEmpty ?? false;

  if (value == null) {
    return value;
  }

  if (isResolvable(value)) {
    const resolved = value.toJSON();
    return resolve(resolved, options);
  }

  // Special resolution for few JavaScript built-in types
  // that by default would be stringified as empty objects ('{}')
  // as they are missing a `toJSON` implementation.
  switch (true) {
    case types.isRegExp(value):
      if (value.flags) {
        throw new Error(
          "RegExp with flags should be explicitly converted to a string"
        );
      }
      return value.source;

    case types.isSet(value):
      return resolve(Array.from(value), options);

    case types.isMap(value):
      return resolve(Object.fromEntries(value), options);

    case types.isBoxedPrimitive(value):
      return resolve(value.valueOf(), options);
  }

  // if value is a function, call it and resolve the result.
  if (typeof value === "function") {
    const resolved = value.apply(undefined, args);
    return resolve(resolved, options);
  }

  if (typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    if (omitEmpty && value.length === 0) {
      return undefined;
    }
    return value.map((x) => resolve(x, options)).filter((x) => x != null); // filter undefined/null/omitted
  }

  // only allow data types (i.e. objects without constructors)
  if (value.constructor && value.constructor?.name !== "Object") {
    throw new Error(
      `only data types can be resolved. trying to resolve object of type ${value.constructor?.name}`
    );
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    const resolved = resolve(v, options);

    // skip undefined values
    if (resolved === undefined) {
      continue;
    }

    result[k] = resolved;
  }

  if (omitEmpty && Object.keys(result).length === 0) {
    return undefined;
  }

  return result;
}
