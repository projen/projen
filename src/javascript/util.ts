import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve, posix } from "path";
import * as semver from "semver";
import { NodePackage } from "./node-package";
import { Project } from "../project";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(posix.sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return posix.join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}

/**
 * Whether the given dependency version is installed
 *
 * This can be used to test for the presence of certain versions of devDependencies,
 * and do something dependency-specific in certain Components. For example, test for
 * a version of Jest and generate different configs based on the Jest version.
 *
 * NOTE: The implementation of this function currently is currently
 * approximate: to do it correctly, we would need a separate implementation
 * for every package manager, to query its installed version (either that, or we
 * would code to query `package-lock.json`, `yarn.lock`, etc...).
 *
 * Instead, we will look at `package.json`, and assume that the versions
 * picked by the package manager match ~that. This will work well enough for
 * major version checks, but may fail for point versions.
 *
 * What we SHOULD do is: `actualVersion ∈ checkRange`.
 *
 * What we do instead is a slightly more sophisticated version of
 * `requestedRange ∩ checkRange != ∅`. This will always give a correct result if
 * `requestedRange ⊆ checkRange`, but may give false positives when `checkRange
 * ⊆ requestedRange`.
 *
 * May return `undefined` if the question cannot be answered. These include the
 * following cases:
 *
 *   - The dependency is requested via local file dependencies (`file://...`)
 *   - The dependency uses an other type of URL, such as a GitHub URL
 *   - The dependency is not found in the `package.json`, such as when
 *     bootstrapping from an external projen package, and the `package.json`
 *     file only has that one external package as a dependency
 *
 * Otherwise it will return `true` if the installed version is probably in the
 * requested range, and `false` if it is probably not.
 *
 * This API may eventually be added to the public projen API, but only after
 * we implement exact version checking.
 *
 * @param dependencyName The name of the dependency
 * @param checkRange A particular version, or range of versions.
 */
export function hasDependencyVersion(
  project: Project,
  dependencyName: string,
  checkRange: string
): boolean | undefined {
  const file = NodePackage.of(project)?.file;
  if (!file) {
    return undefined;
  }

  if (!existsSync(file.absolutePath)) {
    return undefined;
  }

  const pj = JSON.parse(readFileSync(file.absolutePath, "utf-8"));

  // Technicaly, we should be intersecting all ranges to come up with the most narrow dependency
  // range, but `semver` doesn't allow doing that and we don't want to add a dependency on `semver-intersect`.
  //
  // Let's take the first dependency declaration we find, and assume that people
  // set up their `package.json` correctly.
  let requestedRange: string | undefined;
  for (const key of ["dependencies", "devDependencies", "peerDependencies"]) {
    const deps = pj[key] ?? {};
    let requestedVersion = deps[dependencyName];
    if (requestedVersion) {
      // If this is not a valid range, it could be 'file:dep.tgz', or a GitHub URL. No way to know what
      // version we're getting, bail out.
      if (!semver.validRange(requestedVersion)) {
        return undefined;
      }
      requestedRange = requestedVersion;
      break;
    }
  }

  // If the dependency is not found in the `package.json`, we can't answer the question (yet).
  // This means that the dependency hasn't been added yet, which means we know (upstream) what we're going to request,
  // or we're going to ask for '*' and we'll get the latest version.
  if (!requestedRange) {
    return undefined;
  }

  return installedVersionProbablyMatches(requestedRange, checkRange);
}

/**
 * Whether the given requestedRange *probably* leads to the installation of a version that matches checkRange
 *
 * We assume that NPM always installs the most recent version of a package that
 * is allowed by the requestedRange.
 */
export function installedVersionProbablyMatches(
  requestedRange: string,
  checkRange: string
): boolean {
  const options = {
    includePrerelease: true,
    loose: true,
  };

  // No questions asked: always true
  if (semver.subset(requestedRange, checkRange, options)) {
    return true;
  }

  // Also no questions asked: always false
  if (!semver.intersects(requestedRange, checkRange, options)) {
    return false;
  }

  // Now we're in tricky territory. We intersect, but aren't a full subset.
  // We are in one of the following 2 situations, which we will tie-break by
  // assuming NPM will install the most recent matching version in 'requested'.
  //
  // requested  | check    | result
  // -----------|----------|-----------
  //   >= 2     |  >= 3    | probably true (chance of FP)
  //   <= 2     |  <= 1    | probably false (change of FN)
  //
  // `semver` doesn't make it easy to distinguish these cases (we can't request
  // the `maxVersion` that satisfies a range). Instead what we do is
  // get the `minVersion` of each range, and if they compare equal we assume
  // we're in the bottom case with `<=` checks, and return `false`.

  return !semver.eq(
    semver.minVersion(requestedRange, options) ?? "1.2.3",
    semver.minVersion(checkRange, options) ?? "1.2.3"
  );
}
