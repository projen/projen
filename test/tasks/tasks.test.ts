import { Project, TaskRuntime } from "../../src";
import { TasksManifest, TaskStep } from "../../src/task-model";
import { TestProject, synthSnapshot } from "../util";

test("default tasks", () => {
  const p = new TestProject();
  expect(synthTasksManifest(p)).toMatchSnapshot();
});

test("empty task", () => {
  const p = new TestProject();

  // WHEN
  p.addTask("empty");

  // THEN
  expectManifest(p, {
    tasks: {
      empty: {
        name: "empty",
      },
    },
  });
});

test("remove task", () => {
  const p = new TestProject();

  // WHEN
  const task = p.addTask("task1");
  p.addTask("task2");
  const removeTask = p.removeTask("task1");

  // THEN
  expect(removeTask).toEqual(task);
  expectManifest(p, {
    tasks: {
      task2: {
        name: "task2",
      },
    },
  });
});

test("re-add removed task", () => {
  const p = new TestProject();

  // WHEN
  p.addTask("task1");
  p.addTask("task2");
  const removeTask = p.removeTask("task2");
  p.addTask("task2");

  // THEN
  expect(removeTask).toBeTruthy();
  expectManifest(p, {
    tasks: {
      task1: {
        name: "task1",
      },
      task2: {
        name: "task2",
      },
    },
  });
});

test("throw when removing a dependent task", () => {
  const p = new TestProject();

  // WHEN
  const primary = p.addTask("primary");
  const dependent = p.addTask("dependent");
  primary.spawn(dependent);

  // THEN
  expect(() => p.removeTask("dependent")).toThrowError(
    'Unable to remove task "dependent" because the following tasks depend on it: primary'
  );
});

test("remove already removed task", () => {
  const p = new TestProject();

  expect(p.removeTask("task1")).toBe(undefined);
});

test('multiple "exec" commands', () => {
  const p = new TestProject();

  // WHEN
  const task = p.addTask("hello", {
    description: "hello, world",
    exec: "echo hello", // initial command
    env: {
      FOO: "bar",
    },
  });

  task.exec("echo world");
  task.exec('echo "with quotes"');
  task.env("BAR", "baz");

  // THEN
  expectManifest(p, {
    tasks: {
      hello: {
        name: "hello",
        description: "hello, world",
        env: {
          FOO: "bar",
          BAR: "baz",
        },
        steps: [
          { exec: "echo hello" },
          { exec: "echo world" },
          { exec: 'echo "with quotes"' },
        ],
      },
    },
  });
});

test("subtasks", () => {
  // GIVEN
  const p = new TestProject();
  const hello = p.addTask("hello", { exec: "echo hello" });
  const world = p.addTask("world");

  // WHEN
  world.exec('echo "running hello"');
  world.spawn(hello);

  // THEN
  expectManifest(p, {
    tasks: {
      hello: {
        name: "hello",
        steps: [{ exec: "echo hello" }],
      },
      world: {
        name: "world",
        steps: [{ exec: 'echo "running hello"' }, { spawn: "hello" }],
      },
    },
  });
});

test("reset() can be used to reset task steps", () => {
  // GIVEN
  const p = new TestProject();
  const t0 = p.addTask("your-task");
  const t = p.addTask("my-task");
  t.exec("line1");
  t.spawn(t0);
  t.exec("line2");

  // WHEN
  t.reset("line3");
  t.exec("line4", { cwd: "foo" });

  // THEN
  expectManifest(p, {
    tasks: {
      "your-task": {
        name: "your-task",
      },
      "my-task": {
        name: "my-task",
        steps: [{ exec: "line3" }, { cwd: "foo", exec: "line4" }],
      },
    },
  });
});

test("prependXXX() can be used to add steps from the top", () => {
  // GIVEN
  const p = new TestProject();
  const sub = p.addTask("my-sub-task", { exec: "subexec" });

  const t = p.addTask("my-task");
  t.exec("line1");

  // WHEN
  t.prependExec("line2");
  t.prependSpawn(sub);
  t.prependSay("message");

  // THEN
  expectManifest(p, {
    tasks: {
      "my-sub-task": {
        name: "my-sub-task",
        steps: [{ exec: "subexec" }],
      },
      "my-task": {
        name: "my-task",
        steps: [
          { say: "message" },
          { spawn: "my-sub-task" },
          { exec: "line2" },
          { exec: "line1" },
        ],
      },
    },
  });
});

test("env() can be used to add environment variables", () => {
  // GIVEN
  const p = new TestProject();
  const t = p.addTask("my-task", {
    env: {
      INITIAL: "123",
      ENV: "456",
    },
  });

  // WHEN
  t.env("FOO", "BAR");
  t.env("HELLO", "world");

  // THEN
  expectManifest(p, {
    tasks: {
      "my-task": {
        name: "my-task",
        env: {
          INITIAL: "123",
          ENV: "456",
          FOO: "BAR",
          HELLO: "world",
        },
      },
    },
  });
});

test(".envVars returns all environment variables in the task level", () => {
  // GIVEN
  const p = new TestProject();
  const t = p.addTask("my-task", {
    env: {
      INITIAL: "123",
      ENV: "456",
    },
  });

  // WHEN
  t.env("FOO", "BAR");
  t.env("HELLO", "world");

  expect(t.envVars).toStrictEqual({
    INITIAL: "123",
    ENV: "456",
    FOO: "BAR",
    HELLO: "world",
  });
});

test(".steps can be used to list all steps in the current task", () => {
  // GIVEN
  const p = new TestProject();
  const t0 = p.addTask("your");
  const t = p.addTask("my");
  t.exec("step1");
  t.exec("step2");
  t.exec("step3");
  t.spawn(t0);
  t.exec("step4");

  // WHEN
  const steps = t.steps;

  // THEN
  expect(steps).toStrictEqual([
    { exec: "step1" },
    { exec: "step2" },
    { exec: "step3" },
    { spawn: "your" },
    { exec: "step4" },
  ] as TaskStep[]);
});

test("updateStep() can be used to replace a specific step", () => {
  // GIVEN
  const p = new TestProject();
  const t0 = p.addTask("your");
  const t = p.addTask("my");
  t.exec("step1");
  t.exec("step2");
  t.exec("step3");
  t.spawn(t0);
  t.exec("step4");

  // WHEN
  t.updateStep(2, { exec: "edited" });

  // THEN
  expectManifest(p, {
    tasks: {
      your: {
        name: "your",
      },
      my: {
        name: "my",
        steps: [
          { exec: "step1" },
          { exec: "step2" },
          { exec: "edited" },
          { spawn: "your" },
          { exec: "step4" },
        ],
      },
    },
  });
});

test.each([[2], [-2]])(
  "insertStep(%p) can be used to insert a step at a specific location",
  (index) => {
    // GIVEN
    const p = new TestProject();
    const t = p.addTask("my");
    t.exec("step1");
    t.exec("step2");
    t.exec("step3");
    t.exec("step4");

    t.insertStep(index, { exec: "step2.5" });

    // THEN
    expectManifest(p, {
      tasks: {
        my: {
          name: "my",
          steps: [
            { exec: "step1" },
            { exec: "step2" },
            { exec: "step2.5" },
            { exec: "step3" },
            { exec: "step4" },
          ],
        },
      },
    });
  }
);

test("insertStep throws if the index is out of bounds", () => {
  // GIVEN
  const p = new TestProject();
  const t = p.addTask("my");
  t.exec("step1");
  t.exec("step2");
  t.exec("step3");
  t.exec("step4");

  expect(() => t.insertStep(-5, { exec: "asdf" })).toThrow(/out of bounds/);
  expect(() => t.insertStep(5, { exec: "asdf" })).toThrow(/out of bounds/);
});

test("removeStep() can be used to remove a specific step", () => {
  // GIVEN
  const p = new TestProject();
  const t0 = p.addTask("your");
  const t = p.addTask("my");
  t.exec("step1");
  t.exec("step2");
  t.exec("step3");
  t.spawn(t0);
  t.exec("step4");

  // WHEN
  t.removeStep(2);

  // THEN
  expectManifest(p, {
    tasks: {
      your: {
        name: "your",
      },
      my: {
        name: "my",
        steps: [
          { exec: "step1" },
          { exec: "step2" },
          { spawn: "your" },
          { exec: "step4" },
        ],
      },
    },
  });
});

test('"condition" can be used to define a command that will determine if a task should be skipped', () => {
  // GIVEN
  const p = new TestProject();
  p.addTask("foo", {
    condition: "false",
    exec: "foo bar",
  });

  // THEN
  expectManifest(p, {
    tasks: {
      foo: {
        name: "foo",
        condition: "false",
        steps: [{ exec: "foo bar" }],
      },
    },
  });
});

test('"addCondition" can be added after task initialized', () => {
  // GIVEN
  const p = new TestProject();
  const t = p.addTask("foo", {
    condition: undefined,
    exec: "foo bar",
  });
  t.addCondition("false");

  // THEN
  expectManifest(p, {
    tasks: {
      foo: {
        name: "foo",
        condition: "false",
        steps: [{ exec: "foo bar" }],
      },
    },
  });
});

test('"addCondition" can append additional condition', () => {
  // GIVEN
  const p = new TestProject();
  const t = p.addTask("foo", {
    condition: "a",
    exec: "foo bar",
  });
  t.addCondition("b");

  // THEN
  expectManifest(p, {
    tasks: {
      foo: {
        name: "foo",
        condition: "a && b",
        steps: [{ exec: "foo bar" }],
      },
    },
  });
});

test('"builtin" can be used to execute builtin commands', () => {
  const p = new TestProject();
  const task = p.addTask("foo", {
    condition: "false",
  });

  task.builtin("my/builtin");
  task.builtin("your/builtin");

  // THEN
  expectManifest(p, {
    tasks: {
      foo: {
        name: "foo",
        condition: "false",
        steps: [{ builtin: "my/builtin" }, { builtin: "your/builtin" }],
      },
    },
  });
});

test('"requiredEnv" can be used to specify required environment variables', () => {
  const p = new TestProject();
  p.addTask("foo", {
    requiredEnv: ["MISSING1", "MISSING2", "NOT_MISSING"],
  });

  // THEN
  expectManifest(p, {
    tasks: {
      foo: {
        name: "foo",
        requiredEnv: ["MISSING1", "MISSING2", "NOT_MISSING"],
      },
    },
  });
});

test("lock() can be used to disallow modifications", () => {
  const p = new TestProject();
  const t = p.addTask("t1");
  const t2 = p.addTask("t2");
  t.exec("echo hello");
  t.exec("echo world");

  t.lock();

  const error = 'Task "t1" is locked for changes';

  expect(() => t.reset()).toThrow(error);
  expect(() => t.exec("boom")).toThrow(error);
  expect(() => t.prependExec("pre")).toThrow(error);
  expect(() => t.spawn(t2)).toThrow(error);
  expect(() => t.prependSpawn(t2)).toThrow(error);
  expect(() => t.builtin("damn")).toThrow(error);
  expect(() => t.env("Foo", "bar")).toThrow(error);
  expect(() => t.say("hoho")).toThrow(error);
  expect(() => t.prependSay("hoho")).toThrow(error);
});

test("it is possible to edit the description", () => {
  const p = new TestProject();
  const t1 = p.addTask("t1");
  const t2 = p.addTask("t2", { description: "my description" });

  // WHEN
  t1.description = "hello";
  t2.description = "world";

  // THEN
  const files = synthSnapshot(p);
  expect(files[".projen/tasks.json"].tasks.t1.description).toBe("hello");
  expect(files[".projen/tasks.json"].tasks.t2.description).toBe("world");
});

test("steps can receive args", () => {
  const p = new TestProject();

  // WHEN
  const hello = p.addTask("hello");
  const world = p.addTask("world", {
    exec: "echo $@ world",
    receiveArgs: true,
  });

  hello.exec("echo hello", { receiveArgs: true });
  hello.spawn(world, { receiveArgs: true });

  // THEN
  expectManifest(p, {
    tasks: {
      hello: {
        name: "hello",
        steps: [
          { exec: "echo hello", receiveArgs: true },
          { spawn: "world", receiveArgs: true },
        ],
      },
      world: {
        name: "world",
        steps: [{ exec: "echo $@ world", receiveArgs: true }],
      },
    },
  });
});

test("allows setting the cwd for the task", () => {
  const p = new TestProject();
  const t = p.addTask("t", {
    cwd: "foo",
  });

  // WHEN
  t.cwd = "bar";

  // THEN
  expectManifest(p, {
    tasks: {
      t: {
        name: "t",
        cwd: "bar",
      },
    },
  });
});

function expectManifest(p: Project, toStrictEqual: TasksManifest) {
  const manifest = synthTasksManifest(p);
  delete manifest["//"];

  // delete all standard tasks
  delete manifest.tasks.build;
  delete manifest.tasks["pre-compile"];
  delete manifest.tasks.compile;
  delete manifest.tasks["post-compile"];
  delete manifest.tasks.test;
  delete manifest.tasks.default;
  delete manifest.tasks.package;
  delete manifest.tasks.eject;

  expect(manifest).toStrictEqual(toStrictEqual);
}

function synthTasksManifest(p: Project) {
  return synthSnapshot(p)[TaskRuntime.MANIFEST_FILE];
}
