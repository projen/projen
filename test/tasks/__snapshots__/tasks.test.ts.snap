// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`default tasks 1`] = `
{
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
  "tasks": {
    "build": {
      "description": "Full release build",
      "name": "build",
      "steps": [
        {
          "spawn": "default",
        },
        {
          "spawn": "pre-compile",
        },
        {
          "spawn": "compile",
        },
        {
          "spawn": "post-compile",
        },
        {
          "spawn": "test",
        },
        {
          "spawn": "package",
        },
      ],
    },
    "compile": {
      "description": "Only compile",
      "name": "compile",
    },
    "default": {
      "description": "Synthesize project files",
      "name": "default",
    },
    "eject": {
      "description": "Remove projen from the project",
      "env": {
        "PROJEN_EJECTING": "true",
      },
      "name": "eject",
      "steps": [
        {
          "spawn": "default",
        },
      ],
    },
    "package": {
      "description": "Creates the distribution package",
      "name": "package",
    },
    "post-compile": {
      "description": "Runs after successful compilation",
      "name": "post-compile",
    },
    "pre-compile": {
      "description": "Prepare the project for compilation",
      "name": "pre-compile",
    },
    "test": {
      "description": "Run tests",
      "name": "test",
    },
  },
}
`;
